deepspeed/runtime/engine.py

```diff
@@ -125,21 +125,21 @@ def __init__(self,
         if dist_init_required is None:
             dist_init_required = not dist.is_initialized()
 
-        if self._in_aml():
-            self._set_environment_variables_for_nccl_backend(args)
-        else:
-            self._mpi_check(args, dist_init_required)
-
         self.dist_backend = "nccl"
+
         if dist_init_required:
-            if not dist.is_initialized():
-                logger.info("Initializing torch distributed with backend: {}".format(
-                    self.dist_backend))
-                dist.init_process_group(backend=self.dist_backend)
+            # discover using mpi4py if user specifies the flag
+            if hasattr(args, 'deepspeed_mpi') and args.deepspeed_mpi:
+                self._mpi_check(args, dist_init_required)
             else:
-                logger.warning(
-                    "Was given dist_init_required=True but detected that torch"
-                    "distributed was already initialized, cannot initialize twice.")
+                # detect if we are in Azure ML environment
+                if self._in_aml():
```

Comment:
I wonder if we could just do something like the following (see below). This would work I think as long as the ITP/AML environment made sure to init torch distributed before running deepspeed.initialize() which as far as I understand they do?

```python
        if dist_init_required is None:
            dist_init_required = not dist.is_initialized()

        if dist_init_required:
            if self._in_aml():
                self._set_environment_variables_for_nccl_backend(args)
            else:
                self._mpi_check(args, dist_init_required)

            self.dist_backend = "nccl"
            logger.info("Initializing torch distributed with backend: {}".format(self.dist_backend))
            dist.init_process_group(backend=self.dist_backend)
        else:
            assert dist.is_initialized(), 'dist_init_required=False but torch.distributed is not initialized, please init manually or set dist_init_required=True'
```

If for some reason they don't init the backend but they do patch the environment vars we could also add some code that checks to see if all the required environment variables are already patched to get dist.init_process_group(..) to work. If so, we init the backend. (code snippet below comes from #276)

```python
    required_env = ["RANK", "WORLD_SIZE", "MASTER_ADDR", "MASTER_PORT", "LOCAL_RANK"]
    if all(map(lambda v: v in os.environ, required_env)):
        print('safe to init torch distributed')
```

Reply:
But what if there is another user of AML/ITP and they use deepspeed without intializing ddp, then it will throw the same error as Jiacheng saw. I think it is safer to check for the MPI flag first so we don't take the AML path if the user is specifying it already.

Reply:
Maybe let’s clarify Jiacheng’s error. I think they were already initializing ddp but our code currently does the mpi and aml check code even if the backend is initialized which I think is a bug on our side.

Reply:
Agreed. Jiacheng's error will be fixed because we moved the init part inside the if dist_init_required so we are good on that side.